File: cache/cache.go
package cache

import (
	"log"
	"reflect"
	"sync"
	"time"

	"github.com/google/go-cmp/cmp"
	"github.com/miekg/dns"
)

type item struct {
	msg *dns.Msg
	ttl time.Time
	used time.Time
}

type Cache struct {
	sync.Mutex
	items map[string]*item
	size int
	ttl time.Duration
}

func New(capacity int) *Cache {
	cache := new(Cache)
	cache.size = capacity
	cache.items = make(map[string]*item)
	return cache
}

func (cache *Cache) Remove(name string){
	cache.Lock()
	delete(cache.items, name)
	cache.Unlock()
}

func (cache *Cache) Add( name string, m *dns.Msg){	
	if cache.size <= 0 || len(cache.items) + 1 > cache.size{
		cache.Remove(cache.LeastRecentlyUsed())
	}
	
	if _, exists := cache.items[name]; !exists{
		cache.Lock()
		cache.items[name] = &item{m.Copy(), time.Now().UTC().Add(cache.ttl), time.Now()}
		cache.Unlock()
	}

}

func (cache *Cache) Contains(any interface{}) bool{
	switch t := any.(type){
	case string:
		if _, exists := cache.items[t]; exists{
			return true
		}
	case *dns.Msg:
		for _, msg := range cache.items {
			if cmp.Equal(t,msg){
				return true
			}
		}
	default:
		return false
	}	
	return false
}

func (cache *Cache) GetItem(name string) *dns.Msg{
	cache.items[name].used = time.Now()
	return cache.items[name].msg
}

func (cache *Cache) LeastRecentlyUsed() string {
	var LRU string
	var tmpTime time.Time
	for name, msg := range cache.items {
		if len(LRU) == 0 {
			LRU = name
			tmpTime = msg.used
			continue
		}
		if msg.used.Before(tmpTime){
			LRU = name
			tmpTime = msg.used
		}
	}
	return LRU
}
func (cache *Cache) FirstInFirstOut() string {
	if len(cache.items) > 0{
		return reflect.ValueOf(cache.items).MapKeys()[0].String()
	} else {
		return ""
	}
	
}

File: Python Server/Cache/Cache.py

from threading import Lock
import time
from dataclasses import dataclass
import dns.message

@dataclass
class Item:
    msg: dns.message.Message
    ttl: int
    used: int

class Cache:
    def __init__(self, n) -> None:
        self.size = n
        self.items = dict()
        self.ttl =  600000
        self.mutex = Lock()
    
    def remove(self, name):
        self.mutex.acquire()
        del self.items[name]
        self.mutex.release()

    def add(self, name, msg):
        if self.size < 0 or len(self.items) > self.size:
            self.remove(self.LastInFirstOut())
        
        if name not in self.items:
            self.mutex.acquire()
            self.items[name] = Item(msg, time.time()+self.ttl, time.time())
            self.mutex.release()
    
    def Contains(self, any):
        if type(any) == str:
            if any in self.items:
                return True
        if type(any) == dns.message.Message:
            for item in self.items:
                if len(self.items[item].msg.question) > 0 and self.items[item].msg.question[0].name == any.question[0].name:
                    return True
        return False

    def getItem(self, name):
        self.items[name].used = time.time()
        return self.items[name].msg
    
    def LeastRecentlyUsed(self):
        lru = ""
        tmpTime = 0
        for item in self.items:
            if len(lru)==0:
                lru = item
                tmpTime = self.items[item].used
            else:
                if self.items[item].used < tmpTime:
                    lru = item
                    tmpTime = self.items[item].used
            
        return lru

    def FirstInFirstOut(self):
        if len(self.items) > 0:
            return list(self.items)[0]
        else:
            return ""
    
    def LastInFirstOut(self):
        if len(self.items) > 0:
            return list(self.items)[len(self.items)-1]
        else:
            return ""



    
